runtime-sdk/modules/evm/src/lib.rs

```diff
@@ -356,9 +523,75 @@ impl module::MigrationHandler for Module {
     }
 }
 
-impl module::AuthHandler for Module {}
+impl<Cfg: Config> module::AuthHandler for Module<Cfg> {
+    fn authenticate_tx<C: Context>(ctx: &mut C, tx: &Transaction) -> Result<(), CoreError> {
+        // We're only interested in transactions that can be paid with tokens
+        // from the corresponding EVM account.
+        match tx.call.method.as_str() {
+            "evm.Create" | "evm.Call" | "evm.Withdraw" => {}
+            _ => return Err(CoreError::NotAuthenticated),
+        }
+
+        let params = Self::params(ctx.runtime_state());
+        let den = params.token_denomination;
+
+        let evm_acct_addr = Self::derive_caller_from_tx_auth_info(&tx.auth_info)
+            .map_err(|e| CoreError::MalformedTransaction(anyhow::Error::new(e)))?;
+
+        // Check and update nonces on all signer accounts.
+        let mut sdk_accounts_store =
+            storage::PrefixStore::new(ctx.runtime_state(), &modules::accounts::MODULE_NAME);
+        let mut sdk_accounts = storage::TypedStore::new(storage::PrefixStore::new(
+            &mut sdk_accounts_store,
+            &modules::accounts::state::ACCOUNTS,
+        ));
+        for si in tx.auth_info.signer_info.iter() {
+            let addr = si.address_spec.address();
+            let mut acct: modules::accounts::types::Account =
+                sdk_accounts.get(&addr).unwrap_or_default();
+
+            if acct.nonce != si.nonce {
+                return Err(CoreError::InvalidNonce);
+            }
+
+            acct.nonce += 1;
+            sdk_accounts.insert(&addr, acct);
+        }
+
+        // The fee should be set by the user to cover at least:
+        //     gas_price * gas_limit + sdk_fees
+        // And the gas should be set to:
+        //     gas_limit + sdk_gas_limit
+        // The difference between the paid fee and the actually used gas
+        // will be returned in do_evm() above after execution is complete
+        // and the EVM has calculated the actual amount of gas used.
+        let fee = tx.auth_info.fee.amount.amount();
+
+        // Take the tokens from the EVM account and move them into the
+        // fee accumulator.
+        let state = ctx.runtime_state();
+        let store = storage::PrefixStore::new(state, &MODULE_NAME);
+        let mut accounts =
+            storage::TypedStore::new(storage::PrefixStore::new(store, &state::ACCOUNTS));
+        let mut account: evm_backend::Account = accounts.get(&evm_acct_addr).unwrap_or_default();
+
+        if account.balance < fee.into() {
+            return Err(CoreError::InsufficientFeeBalance);
+        }
+
+        account.nonce += 1.into();
```

Comment:
What is the point of these separate EVM nonces given that they are not checked and only SDK nonces are checked? It seems like they could get out of sync when some calls are funded from SDK accounts and others from EVM. And then if someone queries the EVM account nonce to generate a transaction, she may get the wrong nonce? Should we make sure they are synced somehow?

Reply:
I think these are only needed by the code running inside the EVM (in any case, the `evm` crate requires them).  We probably don't need to sync them.  We don't have a way to directly query the EVM nonce, so it shouldn't matter on our side?